### DISCUSSION TITLE
SSR requests to airtable

### DISCUSSION THREAD

**Discussion by Alex-Kostenko** [vercel/next.js#47060] [Help] [answered]
### Summary

I use React (v:"^18.2.0"), next js (v: "^12.3.0"), typescript and airtable (v: "^0.11.5"). 
I have a problem with requests in getStaricProps to airtable.
Requests work well locally, but on the hosting, after changing any data in the airtable, the request returns the old (irrelevant) data, and they are updated only after redeploy.
When loading the page, I send a request and get all the ids and keys (code *1), then after clicking on a certain block, I make a request in useEffect for the id that I already have (code *2). If you change the keys on which the data display is tied, this problem appears with the SSR request. 
I understand that I can move it to useEffect and the problem will disappear, but I need the data to be loaded before the page is rendered.
I have attached code examples. Please advice, am I doing something wrong?

### Additional information

```js
//My getStaticProps (code *1)
export const getStaticProps: GetStaticProps = async ({ locale }) => {
//SSR request to airtable
  const listOfAirTableRows = (await AirTableService.getAllRows()) as ListAirtableRows[];

  if (!listOfAirTableRows) {
    return {
      redirect: {
        destination: '/augument-my-business/serverError',
        permanent: false,
      },
    };
  }

  return {
    props: {
      listOfAirTableRows,
      ...(await serverSideTranslations(locale as string, ['common'])),
    },
  };
};

//request (get all id & keys) 
getAllRows: async (setIsNotification?: (item: boolean) => void) => {
    try {
      const airFields = (await airAPI('Flashcards')
        .select({
          fields: ['key'],
        })
        .all()) as unknown as GetFlashCardsRes[];

      return (
        airFields &&
        airFields.map((res) => {
          return { id: res.id, ...res.fields };
        })
      );
    } catch (e) {
      if ((e as IError).statusCode && (e as IError).statusCode > 399 && setIsNotification) {
        console.log('Error in getAllRows');
        setIsNotification(true);
      }
    }
  }

//My request in useEffect to get data by id (code *2)
 useEffect(() => {
   const elemsById = listOfAirTableRows
      .map((item: ListAirtableRows) => {
        if (item.key && item.key.find((value: string) => value === activeCase.activeCell?.key)) {
          return item;
        }
        return null;
      })
      .filter(Boolean) as ListAirtableRows;
  }, [activeCase.activeCell?.key, listOfAirTableRows, activeCase, prevPageData]);
```

**Comment by icyJoseph** [comment, marked as answer]
Have you looked into using `ISR`?

- https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration
- https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation

This option would just rebuild the page after a given time, revalidate prop:

```js
// This function gets called at build time on server-side.
// It may be called again, on a serverless function, if
// revalidation is enabled and a new request comes in
export async function getStaticProps() {
  const res = await fetch('https://.../posts')
  const posts = await res.json()

  return {
    props: {
      posts,
    },
    // Next.js will attempt to re-generate the page:
    // - When a request comes in
    // - At most once every 10 seconds
    revalidate: 10, // In seconds
  }
}
```

**Reply by Alex-Kostenko** [reply]
That helped! But all my local translations are broken on this page now.

**Reply by Alex-Kostenko** [reply]
Could you tell me what could happen with i18-next translations after adding revalidate to getStaticProps?

**Reply by icyJoseph** [reply]
Mmm I have no idea...

Do you have something like this?

```js
loadTranslations()

const Page = () => {}

export async function getStaticProps() {}

export default Page
```

I wonder if when revalidate happens, the `loadTranslations` bit is not included in the server bundle... 

- https://github.com/i18next/next-i18next/issues/1577

**Reply by Alex-Kostenko** [reply]
I have only page like this:

const page: FC<IAirtableRows> = ({ listOfAirTableRows }) => {
const { t, i18n } = useTranslation('common');
 }
export default page;

getStaticProps...

**Reply by Alex-Kostenko** [reply]
<img width="519" alt="Screenshot 2023-03-13 at 13 47 41" src="https://user-images.githubusercontent.com/28589064/224693477-23b15865-54a6-4404-9300-45841926b26f.png">

**Reply by Alex-Kostenko** [reply]
Added localePath: path.resolve('public/locales'), to next-config. Everything works now. Thank you for your attention!
